README.md
# 371pass

___
You should use this README to document things you did that you think are clever and could get you marks, or things you
couldn't get working and perhaps why. It is fully expected that a lot of students will not complete this coursework—it
is designed to be challenging. The mark scheme is also designed to be fair—marks are awarded for everything you
complete, with some given for the quality of the code you write.
___
All required functionality specified in the specification is present and working fully as far as my own and the
automated tests can show, including functionality mentioned in the coursework discussion question board such as
favouring the entries held in a new container over the existing container and being able to update multiple containers
at once (e.g. at the category and item simultaneously).

All functions are commented above the provided comments (e.g. "my comment" followed by "provided comment"), except those
that are trivial and the code is self documenting, i.e. the size() and empty() member functions in the container
classes, and the provided functions are in the same order as provided, but may have some new member functions between
them, such as in the 371pass class where each action has been extracted and placed in order after the run member
function to improve the readability of the code, or in the container classes where the merge function follows the
addItem and addCategory functions.

All variable names are written in snake_case and all member functions names are written in camelCase

CHANGELOG
// -----------------------------------------------------
// CSC371 Advanced Object Oriented Programming (2021/22)
// Department of Computer Science, Swansea University
//
// Author: Martin Porcheron
//         m.a.w.porcheron@swansea.ac.uk
//
// Canvas: https://canvas.swansea.ac.uk/courses/24793
// -----------------------------------------------------
// 371pass Coursework — Changelog
// -----------------------------------------------------


Releases
--------
v1.0.   Update test2.cpp                        3rd Mar 2022
v1.0.1  Minor fixes to framework and worksheet  3rd Mar 2022
v1.0    Initial coursework release              28th Feb 2022


v1.1 - 3rd Mar 2022
-------------------
Update test2.cpp to correctly not expect an exception (Thanks Samuel Roach!)

v1.0.1 — 3rd Mar 2022
---------------------
Minor fixes to framework and worksheet:
 - Clarify student ID only needs to be in editable .cpp/.h files
 - Update Catch2 framework
 - Remove erroneous semicolon from 371pass.h
 - Fix building with build.bat (Windows only)

v1.0 — 28th Feb 2022
--------------------
Coursework released



src/item.h
// -----------------------------------------------------
// CSC371 Advanced Object Oriented Programming (2021/22)
// Department of Computer Science, Swansea University
//
// Author: 851784
//
// Canvas: https://canvas.swansea.ac.uk/courses/24793
// -----------------------------------------------------
// An Item class contains multiple 'entries' as
// key/value pairs (e.g., a key might be 'username'
// and a value would be the username, another might be
// 'url' and the value is the website address the
// username is for.
// -----------------------------------------------------

#ifndef ITEM_H
#define ITEM_H

#include <string>
#include "lib_json.hpp"

class Item {
    std::string identifier;
    std::map<std::string, std::string> entries;
public:
    Item(const std::string identifier);

    unsigned int size();

    bool empty();

    std::string getIdent();

    void setIdent(std::string identifier);

    bool addEntry(std::string key, std::string value);

    void mergeEntries(Item &other);

    std::string getEntry(std::string key);

    bool deleteEntry(std::string key);

    friend bool operator==(const Item &lhs, const Item &rhs);

    std::string str() const;

};

#endif // ITEM_H


src/item.cpp
// -----------------------------------------------------
// CSC371 Advanced Object Oriented Programming (2021/22)
// Department of Computer Science, Swansea University
//
// Author: 851784
//
// Canvas: https://canvas.swansea.ac.uk/courses/24793
// -----------------------------------------------------

#include <sstream>
#include "item.h"

/* Example:
  Item iObj{"identIdent"};*/
Item::Item(const std::string identifier) : identifier(identifier) {}

unsigned int Item::size() {
    return entries.size();
}

bool Item::empty() {
    return entries.empty();
}


/* Example:
  Item iObj{"identIdent"};
  auto ident = iObj.getIdent();*/
std::string Item::getIdent() {
    return identifier;
}

/* Example:
  Item iObj{"identIdent"};
  iObj.setIdent("identIdent2");*/
void Item::setIdent(const std::string identifier) {
    this->identifier = identifier;
}

/* Example:
  Item iObj{"identIdent"};
  iObj.addEntry("key", "value");*/
bool Item::addEntry(const std::string key, const std::string value) {
    auto search = entries.find(key);
    if (search != entries.end()) {
        entries.insert(std::make_pair(key, value));
        return false;
    } else {
        entries.insert(std::make_pair(key, value));
        return true;
    }
}

void Item::mergeEntries(Item &other) {
    entries.insert(other.entries.begin(), other.entries.end());
}

/* Example:
  Item iObj{"identIdent"};
  iObj.addEntry("key", "value");
  auto value = iObj.getEntry("key");*/
std::string Item::getEntry(const std::string key) {
    auto search = entries.find(key);
    if (search != entries.end()) {
        return search->second;
    } else {
        throw std::out_of_range("No such key found.");
    }
}

/* Example:
  Item iObj{"identIdent"};
  iObj.addEntry("key", "value");
  iObj.deleteEntry("key");*/
bool Item::deleteEntry(const std::string key) {
    auto search = entries.find(key);
    if (search != entries.end()) {
        entries.erase(key);
        if (entries.find(key) == entries.end()) return true;
    }
    throw std::out_of_range("Error deleting entry.");
}


/* Example:
  Item iObj1{"identIdent"};
  iObj1.addEntry("key", "value");
  Item iObj2{"identIdent2"};
  if(iObj1 == iObj2) {
    ...
  }*/
bool operator==(const Item &lhs, const Item &rhs) {
    return lhs.identifier == rhs.identifier && lhs.entries == rhs.entries;
}

/* See the coursework specification for how this JSON should look.

 Example:
  Item iObj{"itemIdent"};
  std::string s = iObj.str();*/
std::string Item::str() const {
    std::stringstream sstr;
    sstr << "\"" << identifier << "\":{";
    for (auto it = entries.begin(); it != entries.end(); it++) {
        sstr << "\"" << it->first << "\":\"" << it->second << "\"";
        if (std::next(it) != entries.end()) sstr << ",";
    }
    sstr << "}";
    return sstr.str();
}

src/wallet.h
// -----------------------------------------------------
// CSC371 Advanced Object Oriented Programming (2021/22)
// Department of Computer Science, Swansea University
//
// Author: 851784
//
// Canvas: https://canvas.swansea.ac.uk/courses/24793
// -----------------------------------------------------
// The root object that holds all data for 371pass. This
// class contains Categories, which in turn hold Items,
// which hold a mapping of entries.
// -----------------------------------------------------

#ifndef WALLET_H
#define WALLET_H

#include "category.h"

class Wallet {
    std::map<std::string, Category> categories;
public:
    Wallet();

    unsigned int size();

    bool empty();

    Category newCategory(std::string category_identifier);

    bool addCategory(Category category);

    Category getCategory(std::string category_identifier);

    bool deleteCategory(const std::string category_identifier);

    bool load(std::string filename);

    bool save(std::string filename);

    friend bool operator==(const Wallet &lhs, const Wallet &rhs);

    std::string str() const;
};

#endif // WALLET_H


src/371pass.h
// -----------------------------------------------------
// CSC371 Advanced Object Oriented Programming (2021/22)
// Department of Computer Science, Swansea University
//
// Author: 851784
//
// Canvas: https://canvas.swansea.ac.uk/courses/24793
// -----------------------------------------------------
// This file contains declarations for the helper
// functions used for initialising and running 371pass.
// -----------------------------------------------------

#ifndef _371PASS_H
#define _371PASS_H

#include <string>

#include "lib_cxxopts.hpp"
#include "wallet.h"

namespace App {

// TODO: Enter your student number here! - DONE!
    const std::string STUDENT_NUMBER = "851784";

// Enums (short for enumerations) are similar to their Java implementation.
// It is a user-defined type, used to assign names to internal constants
// in code, instead of simply passing in integers/strings etc.
//
// For example, functions can take a value/constant from a specific enum
// and use it in a switch statement, doing different things
//
// As such, it is a useful way for us to specify information in both a
// machine and human-readable format.
//
// Unlike Java, enums in C++ only map to integer values. You can either let
// the compiler generate the values automatically, in which it allocates a
// unique integer (0-indexed). Or, you can set the value by giving the name
// followed by = <value> (e.g. CREATE=0).
//
// This enum specifies the four different values we support in the action
// program argument.
    enum Action {
        CREATE, READ, UPDATE, DELETE
    };

    int run(int argc, char *argv[]);

    cxxopts::Options cxxoptsSetup();

    App::Action parseActionArgument(cxxopts::ParseResult &args);

    std::string getJSON(Wallet &w);

    std::string getJSON(Wallet &w, const std::string &c);

    std::string getJSON(Wallet &w, const std::string &c, const std::string &i);

    std::string getJSON(Wallet &wObj, const std::string &c, const std::string &i,
                        const std::string &e);

} // namespace App

#endif // _371PASS_H

src/category.h
// -----------------------------------------------------
// CSC371 Advanced Object Oriented Programming (2021/22)
// Department of Computer Science, Swansea University
//
// Author: 851784
//
// Canvas: https://canvas.swansea.ac.uk/courses/24793
// -----------------------------------------------------
// A category contains one or more Items, each with
// their own identifier ('ident').
// -----------------------------------------------------

#ifndef CATEGORY_H
#define CATEGORY_H

#include "item.h"

class Category {
    std::string identifier;
    std::map<std::string, Item> items;
public:
    Category(const std::string identifier);

    unsigned int size();

    bool empty();

    std::string getIdent();

    void setIdent(std::string identifier);

    Item newItem(const std::string item_identifier);

    bool addItem(Item item);

    void mergeItems(Category other);

    Item &getItem(const std::string item_identifier);

    bool deleteItem(const std::string item_identifier);

    friend bool operator==(const Category &lhs, const Category &rhs);

    std::string str() const;

};
#endif // CATEGORY_H


src/wallet.cpp
// -----------------------------------------------------
// CSC371 Advanced Object Oriented Programming (2021/22)
// Department of Computer Science, Swansea University
//
// Author: 851784
//
// Canvas: https://canvas.swansea.ac.uk/courses/24793
// -----------------------------------------------------

#include <fstream>
#include <iostream>
#include <sstream>
#include "lib_json.hpp"
#include "wallet.h"

/* Example:
  Wallet wObj{};*/

Wallet::Wallet() {}

/* Example:
  Wallet wObj{};
  auto size = wObj.size();*/

unsigned int Wallet::size() {
    return categories.size();
}

/* Example:
  Wallet wwObj{};
  auto isEmpty = wObj.empty();*/

bool Wallet::empty() {
    return categories.empty();
}

/* Example:
  Wallet wObj{};
  wObj.newCategory("categoryIdent");*/

Category Wallet::newCategory(std::string category_identifier) {
    auto search = categories.find(category_identifier);
    if (search == categories.end()) {
        Category new_category(category_identifier);
        categories.insert(std::make_pair(category_identifier, new_category));

        if (categories.find(category_identifier) == categories.end()) {
            throw std::runtime_error("Category was not inserted successfully.");
        }

        Category &ref = new_category;
        return ref;
    } else {
        return categories.find(category_identifier)->second;
    }
}

/* Example:
  Wallet wObj{};
  Category cObj{"categoryIdent"};
  wObj.addCategory(cObj);*/

bool Wallet::addCategory(Category category) {
    auto search = categories.find(category.getIdent());
    if (search == categories.end()) {
        categories.insert(std::make_pair(category.getIdent(), category));

        if (categories.find(category.getIdent()) == categories.end()) {
            throw std::runtime_error("Item was not inserted successfully.");
        }

        return true;
    } else {
        categories.find(category.getIdent())->second.mergeItems(category);
        return false;
    }
}

/* Example:
 Wallet wObj{};
  wObj.newCategory("categoryIdent");
  auto cObj = wObj.getCategory("categoryIdent");*/

Category Wallet::getCategory(std::string category_identifier) {
    auto search = categories.find(category_identifier);
    if (search != categories.end()) {
        Category &ref = search->second;
        return ref;
    } else {
        throw std::out_of_range("No such item found.");
    }
}


/* Example:
  Wallet wObj{};
  wObj.newCategory("categoryIdent");
  wObj.deleteCategory("categoryIdent");*/

bool Wallet::deleteCategory(const std::string category_identifier) {
    auto search = categories.find(category_identifier);
    if (search != categories.end()) {
        categories.erase(category_identifier);
        if (categories.find(category_identifier) == categories.end()) return true;
    }
    throw std::out_of_range("No such item found.");
}


/* A note on clashes:
  If you encounter two categories with the same key, the categories should be
  merged (not replaced!). If you encounter two items with the same key in the
  same category, the items should be merged (not replaced!). If you encounter
  two entries with the same key in the same item, the entries should be merged
  (undefined as to which value is picked). Two items in different categories
  can have the same key, as can two entries in different items.

 JSON formatting:
  The JSON file has the following format (see the sample database.json file
  also provided with the coursework framework):
    {
      "Category 1" : {
        "Item 1":  {
          "detail 1 key": "detail 1 value",
          "detail 2 key": "detail 2 value",
          ...
        },
        "Item 2":  {
          "detail 1 key": "detail 1 value",
          "detail 2 key": "detail 2 value"
        },
        ...
      },
      "Category 2": {
        "Item 1": {
          "detail 1 key": "detail 1 value"
        }
        ...
      }
    }

 More help:
  To help you with this function, I've selected the nlohmann::json
  library that you must use for your coursework. Read up on how to use it
  here: https://github.com/nlohmann/json. You may not use any other external
  library other than the one I have provided. You may choose to process the
  JSON yourself without the help of the library but I guarantee this will be
  more work.

  Understanding how to use external libraries is part of this coursework! You
  will need to use this file to deserialize the JSON from string
  to the JSON object provided by this library. Don't just look at the code
  provided below, or in the README on the GitHub and despair. Google search,
  look around, try code out in a separate file to all figure out how to use
  this library.

  Once you have deserialized the JSON string into an object, you will need to
  loop through this object, constructing Category and Item objects according
  to the JSON data in the file.

 Example:
  Wallet wObj{};
  wObj.load("database.json");*/

bool Wallet::load(std::string filename) {
    std::fstream json_file(filename);
    if (json_file.is_open()) {
        std::string json_str((std::istreambuf_iterator<char>(json_file)),
                             (std::istreambuf_iterator<char>()));

        auto json_obj = nlohmann::json::parse(json_str);

        for (auto cat_it = json_obj.begin(); cat_it != json_obj.end(); cat_it++) {
            Category new_category(cat_it.key());
            auto category_items = cat_it.value();

            for (auto item_it = category_items.begin(); item_it != category_items.end(); item_it++) {
                Item new_item(item_it.key());
                auto item_entries = item_it.value();

                for (auto entry_it = item_entries.begin(); entry_it != item_entries.end(); entry_it++) {
                    new_item.addEntry(entry_it.key(), entry_it.value());
                }

                try {
                    Item existing_item = new_category.getItem(item_it.key());
                    existing_item.mergeEntries(new_item);
                } catch (std::out_of_range &e) {
                    new_category.addItem(new_item);
                }
            }

            try {
                Category existing_category = getCategory(cat_it.key());
                existing_category.mergeItems(new_category);
            } catch (std::out_of_range &e) {
                addCategory(new_category);
            }
        }
        return true;
    }
    throw std::runtime_error("Json file did not open successfully");
}

/* Example:
  Wallet wObj{};
   wObj.load("database.json");
  ...
  wObj.save("database.json");*/

bool Wallet::save(std::string filename) {
    std::fstream json_file(filename);
    if (json_file.is_open()) {
        json_file << str();
        return true;
    }
    throw std::runtime_error("Json file did not open successfully");
}

/* Example:
  Wallet wObj1{};
  Wallet wObj2{};
  if(wObj1 == wObj2) {
    ...
  }*/
bool operator==(const Wallet &lhs, const Wallet &rhs) {
    return lhs.categories == rhs.categories;
}

/* Hint:
  See the coursework specification for how this JSON should look.

 Example:
  Wallet wObj{};
  std::string s = wObj.str();*/

std::string Wallet::str() const {
    std::stringstream sstr;
    sstr << "{";
    for (auto it = categories.begin(); it != categories.end(); it++) {
        sstr << it->second.str();
        if (std::next(it) != categories.end()) sstr << ",";
    }
    sstr << "}";
    return sstr.str();
}


src/category.cpp
// -----------------------------------------------------
// CSC371 Advanced Object Oriented Programming (2021/22)
// Department of Computer Science, Swansea University
//
// Author: 851784
//
// Canvas: https://canvas.swansea.ac.uk/courses/24793
// -----------------------------------------------------

#include <sstream>
#include "category.h"


/* Example:
  Category c{"categoryIdent"};*/

Category::Category(std::string identifier) : identifier(identifier) {}

unsigned int Category::size() {
    return items.size();
}

bool Category::empty() {
    return items.empty();
}

/* Example:
  Category cObj{"categoryIdent"};
  auto ident = cObj.getIdent();*/

std::string Category::getIdent() {
    return identifier;
}

/*Example:
 Category cObj{"categoryIdent"};
 cObj.setIdent("categoryIdent2");*/

void Category::setIdent(const std::string identifier) {
    this->identifier = identifier;
}

/* Example:
  Category cObj{"categoryIdent"};
  cObj.newItem("itemIdent");*/

Item Category::newItem(const std::string item_identifier) {
    auto search = items.find(item_identifier);
    if (search == items.end()) {
        Item new_item(item_identifier);
        items.insert(std::make_pair(item_identifier, new_item));

        if (items.find(item_identifier) == items.end()) {
            throw std::runtime_error("Item was not inserted successfully.");
        }

        Item &ref = new_item;
        return ref;
    } else {
        return items.find(item_identifier)->second;
    }
}

/* Example:
  Category cObj{"categoryIdent"};
  Item iObj{"itemIdent"};
  cObj.addItem(iObj);*/

bool Category::addItem(Item item) {
    auto search = items.find(item.getIdent());
    if (search == items.end()) {
        items.insert(std::make_pair(item.getIdent(), item));

        if (items.find(item.getIdent()) == items.end()) {
            throw std::runtime_error("Item was not inserted successfully.");
        }

        return true;
    } else {
        items.find(item.getIdent())->second.mergeEntries(item);
        return false;
    }
}

void Category::mergeItems(Category other) {
    items.insert(other.items.begin(), other.items.end());
}


/* Hint:
  See the test scripts for the exception expected.

 Example:
  Category cObj{"categoryIdent"};
  cObj.newItem("itemIdent");
  auto iObj = cObj.getItem("itemIdent");*/

Item &Category::getItem(const std::string item_identifier) {
    auto search = items.find(item_identifier);
    if (search != items.end()) {
        Item &ref = search->second;
        return ref;
    } else {
        throw std::out_of_range("No such item found.");
    }
}

/* Example:
  Category cObj{"categoryIdent"};
  cObj.newItem("itemIdent");
  bool result = cObj.deleteItem("itemIdent");*/

bool Category::deleteItem(const std::string item_identifier) {
    auto search = items.find(item_identifier);
    if (search != items.end()) {
        items.erase(item_identifier);
        if(items.find(item_identifier) == items.end()) return true;
    }
    throw std::out_of_range("No such item found.");
}

/* Example:
  Category cObj1{"categoryIdent1"};
  cObj1.newItem("itemIdent");
  Category cObj2{"categoryIdent2"};
  if(cObj1 == cObj2) {
    ...
*/
bool operator==(const Category &lhs, const Category &rhs) {
    return lhs.identifier == rhs.identifier && lhs.items == rhs.items;
}


/* See the coursework specification for how this JSON should look.

 Example:
  Category cObj{"categoryIdent"};
  std::string s = cObj.str();*/
std::string Category::str() const {
    std::stringstream sstr;
    sstr << "\"" << identifier << "\":{";
    for (auto it = items.begin(); it != items.end(); it++) {
        sstr << it->second.str() << "";
        if (std::next(it) != items.end()) sstr << ",";
    }
    sstr << "}";
    return sstr.str();
}

src/371pass.cpp
// -----------------------------------------------------
// CSC371 Advanced Object Oriented Programming (2021/22)
// Department of Computer Science, Swansea University
//
// Author: 851784
//
// Canvas: https://canvas.swansea.ac.uk/courses/24793
// -----------------------------------------------------

#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>

#include "371pass.h"
#include "lib_cxxopts.hpp"
#include "wallet.h"

// TODO Complete this function. You have been provided some skeleton code which
//  retrieves the database file name from cxxopts.
//  1. Load the database file by calling load() on a Wallet object
//  2. Parse the 'action' argument to decide what action should be taken (
//     create, read, update, or delete). Read is the easiest to implement, and
//     update is the hardest. The details of how these arguments work is in the
//     coursework specification.
//  4. Save the updated Wallet object to the JSON file if there have been
//     changes (calling save() on the Wallet object).
//
/* Some commented out code has been provided. Using some of this will be
 demonstrated in the coursework video on Canvas. Remember, this coursework is
 meant to be challenging and testing your understanding of programming in C++.
 Part of the challenge is figuring things out on your own. That is a major
 part of software development.

 Example:
  int main(int argc, char *argv[]) { return App::run(argc, argv); }*/
int App::run(int argc, char *argv[]) {
    auto options = App::cxxoptsSetup();
    auto args = options.parse(argc, argv);
    args.arguments().size();

    // Print the help usage if requested
    if (args.count("help")) {
        std::cout << options.help() << '\n';
        return 0;
    }

    // Open the database and construct the Wallet
    const std::string db = args["db"].as<std::string>();
    Wallet wObj{};
    wObj.load(db);

    const Action a = parseActionArgument(args);
    try {
        switch (a) {
            case Action::CREATE:
                if (args["category"].count() > 0 && args["item"].count() > 0
                    && args["entry"].count() > 0) {
                    std::string entry_pair = args["entry"].as<std::string>();
                    std::string entry_delimiter = ",";
                    if (entry_pair.find(entry_delimiter)) {
                        std::string entry_identifier = entry_pair.substr(0, entry_pair.find(entry_delimiter));
                        std::string entry_value = entry_pair.substr(entry_pair.find(entry_delimiter) + 1,
                                                                    entry_pair.length());
                    }
                }
                break;

            case Action::READ:
                if (args["category"].count() > 0 && args["item"].count() > 0 && args["entry"].count() > 0) {
                    std::string value = wObj.getCategory(args["category"].as<std::string>())
                            .getItem(args["item"].as<std::string>())
                            .getEntry(args["entry"].as<std::string>());
                    std::cout << value;
                } else if (args["category"].count() > 0 && args["item"].count() > 0) {
                    std::cout << getJSON(wObj, args["category"].as<std::string>(), args["item"].as<std::string>());
                } else if (args["category"].count() > 0) {
                    Category cat = wObj.getCategory(args["category"].as<std::string>());
                    std::cout << getJSON(wObj, args["category"].as<std::string>());
                } else {
                    std::cout << getJSON(wObj);
                }
                break;
            case Action::UPDATE:
                throw std::runtime_error("update not implemented");
                break;

            case Action::DELETE:
                throw std::runtime_error("delete not implemented");
                break;

            default:
                throw std::runtime_error("Unknown action not implemented");
        }
        return 0;
    } catch (std::out_of_range &e) {
        std::cerr << e.what();
        return 1;
    }


    return 0;
}

/* Create a cxxopts instance. You do not need to modify this function.

 Example:
  auto options = App::cxxoptsSetup();
  auto args = options.parse(argc, argv);*/
cxxopts::Options App::cxxoptsSetup() {
    cxxopts::Options cxxopts("371pass", "Student ID: " + STUDENT_NUMBER + "\n");

    cxxopts.add_options()(
            "db", "Filename of the 371pass database",
            cxxopts::value<std::string>()->default_value("database.json"))(

            "action", "Action to take, can be: 'create', 'read', 'update', 'delete'.",
            cxxopts::value<std::string>())(

            "category",
            "Apply action to a category (e.g., if you want to add a category, set the"
            " action argument to 'add' and the category argument to your chosen"
            " category identifier).",
            cxxopts::value<std::string>())(

            "item",
            "Apply action to an item (e.g., if you want to add an item, set the "
            "action argument to 'add', the category argument to your chosen category "
            "identifier and the item argument to the item identifier).",
            cxxopts::value<std::string>())(

            "entry",
            "Apply action to an entry (e.g., if you want to add an entry, set the "
            "action argument to 'add', the category argument to your chosen category "
            "identifier, the item argument to your chosen item identifier, and the "
            "entry argument to the string 'key,value'). If there is no comma, an "
            "empty entry is inserted. If you are simply retrieving an entry, set the "
            "entry argument to the 'key'. If you are updating an entry key, use a : "
            "e.g., oldkey:newkey,newvalue.",
            cxxopts::value<std::string>())(

            "h,help", "Print usage.");

    return cxxopts;
}

/* Example:
  auto options = App::cxxoptsSetup();
  auto args = options.parse(argc, argv);
  App::Action action = parseActionArgument(args);*/
App::Action App::parseActionArgument(cxxopts::ParseResult &args) {
    std::string input = args["action"].as<std::string>();
    std::transform(input.begin(), input.end(), input.begin(), ::toupper);

    if (input == "CREATE") {
        return Action::CREATE;
    }
    if (input == "READ") {
        return Action::READ;
    }
    if (input == "UPDATE") {
        return Action::UPDATE;
    }
    if (input == "DELETE") {
        return Action::DELETE;
    }
    throw std::invalid_argument("action");
}


/* This function has been implemented for you, but you may wish to adjust it.
 You will have to uncomment the code, which has been left commented to ensure
 the coursework framework compiles (i.e., it calls functions that you must
 implement, once you have implemented them you may uncomment this function).

 Example:
  Wallet wObj{};
  std::cout << getJSON(wObj);*/
std::string App::getJSON(Wallet &wObj) {
    return wObj.str();
}

/* This function has been implemented for you, but you may wish to adjust it.
 You will have to uncomment the code, which has been left commented to ensure
 the coursework framework compiles (i.e., it calls functions that you must
 implement, once you have implemented them you may uncomment this function).

 Example:
  Wallet wObj{};
  std::string c = "category argument value";
  std::cout << getJSON(wObj, c);*/
std::string App::getJSON(Wallet &wObj, const std::string &c) {
    auto cObj = wObj.getCategory(c);
    return cObj.str();
}

/* This function has been implemented for you, but you may wish to adjust it.
 You will have to uncomment the code, which has been left commented to ensure
 the coursework framework compiles (i.e., it calls functions that you must
 implement, once you have implemented them you may uncomment this function).

 Example:
  Wallet wObj{};
  std::string c = "category argument value";
  std::string i = "item argument value";
  std::cout << getJSON(wObj, c, i);*/
std::string App::getJSON(Wallet &wObj, const std::string &c,
                         const std::string &i) {
    auto cObj = wObj.getCategory(c);
    const auto iObj = cObj.getItem(i);
    return iObj.str();
}


/* This function has been implemented for you, but you may wish to adjust it.
 You will have to uncomment the code, which has been left commented to ensure
 the coursework framework compiles (i.e., it calls functions that you must
 implement, once you have implemented them you may uncomment this function).

 Example:
  Wallet wObj{};
  std::string c = "category argument value";
  std::string i = "item argument value";
  std::string e = "entry argument value";
  std::cout << getJSON(wObj, c, i, e);*/
std::string App::getJSON(Wallet &wObj, const std::string &c,
                         const std::string &i, const std::string &e) {
    // Only uncomment this once you have implemented the functions used!
    auto cObj = wObj.getCategory(c);
    auto iObj = cObj.getItem(i);
    return iObj.getEntry(e);
}